### PR TITLE
feature(form): add several form validators and error messages 

### PR DESCRIPTION
**Pull request by nappynapster**
@plentymarkets/team-terra

#### Must be done by Terra

Testing

-   [ ] Person 1 (mandatory)

    Browser support (relevant for changes in scss / appearance of components)

    -   [ ] Chrome
    -   [ ] Firefox
    -   [ ] Safari

    Responsiveness

    -   [ ] Desktop
    -   [ ] Tablet
    -   [ ] Mobile

-   [ ] Person 2 (mandatory)

    Browser support

    -   [ ] Chrome
    -   [ ] Firefox
    -   [ ] Safari

    Responsiveness

    -   [ ] Desktop
    -   [ ] Tablet
    -   [ ] Mobile

Information transfer

-   [ ] Inform Ceres about changes in `TerraFormComponent` (optional: relevant for changes in this component)

Terra Basic Plugin

-   [ ] Adapt changes (optional: relevant for version updates, global design changes, l10n)

---

#### Must be done by every developer

Please inform a member of Terra (@plentymarkets/team-terra) to get the upper part of the checklist done (with urgency or deadline).

Documentation

-   [ ] Changelog (optional: relevant for every change which influences the API)
-   [ ] Example (updated or created)
-   [ ] JSDoc (optional: relevant for every change which influences the API)
-   [ ] [Google spreadsheet](https://docs.google.com/spreadsheets/d/1OINnux8TEoitV-qdAxqUQaf9oGI_fqFVwfRWenRFfhI/edit#gid=0) (relevant when deprecating public APIs or features)

Testing

-   [ ] Unit Test (updated or created)

### CODE REVIEW

File: src/lib/components/forms/form/form-components/double-input/double-input.component.html [hunk 1]
```diff
@@ -12,5 +12,13 @@
         [style.text-align]="isPriceInput ? 'right' : 'left'"
         [pattern]="_regex"
         [step]="_step"
+        #control="ngModel"
     />
+    <mat-error *ngIf="isRequired && control.hasError('required')">
+        {{ 'validators.required' | translate: _locale.language }}
+    </mat-error>
+
```
Comment (niwolf) [inline src/lib/components/forms/form/form-components/double-input/double-input.component.html:20]:
```suggestion
```

**Review by niwolf — COMMENTED** [review]
ansonsten 👍  & tested

**Review by niwolf — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,14 @@
+# 11.5.0 (XX.XX.2021)
+
+### Features
+
+-   **terra-form**: error messages have been added for the following form field validators:
```
Comment (johannabecker) [inline CHANGELOG.md:5]:
```suggestion
-   **terra-form**: Error messages have been added for the following form field validators:
```

Comment (johannabecker) [inline CHANGELOG.md:5]:
```suggestion
-   **terra-form**: Error messages have been added for the following form field validators:
```

**Review by johannabecker — COMMENTED** [review]
Nur eine Kleinigkeit. Changelog sieht sonst gut aus.

File: src/lib/components/forms/form/form-components/datepicker/datepicker.component.html [hunk 3]
```diff
@@ -8,7 +8,11 @@
         [(ngModel)]="value"
         (ngModelChange)="_onChange($event)"
         [required]="isRequired"
+        #control="ngModel"
     />
     <mat-datepicker-toggle matSuffix [for]="picker"></mat-datepicker-toggle>
     <mat-datepicker #picker></mat-datepicker>
+    <mat-error *ngIf="isRequired && control.hasError('required')">
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/datepicker/datepicker.component.html:15]: s.o.
```suggestion
    <mat-error *ngIf="control.hasError('required')">
```

File: src/lib/components/forms/form/form-components/double-input/double-input.component.html [hunk 4]
```diff
@@ -12,5 +12,12 @@
         [style.text-align]="isPriceInput ? 'right' : 'left'"
         [pattern]="_regex"
         [step]="_step"
+        #control="ngModel"
     />
+    <mat-error *ngIf="isRequired && control.hasError('required')">
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/double-input/double-input.component.html:17]:
```suggestion
    <mat-error *ngIf="control.hasError('required')">
```

File: src/lib/components/forms/form/form-components/multi-select/multi-select.component.html [hunk 5]
```diff
@@ -23,4 +24,7 @@
             {{ item.caption }}
         </mat-option>
     </mat-select>
+    <mat-error *ngIf="isRequired && control.hasError('required')">
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/multi-select/multi-select.component.html:27]:
```suggestion
    <mat-error *ngIf="control.hasError('required')">
```

File: src/lib/components/forms/form/form-components/number-input/number-input.component.html [hunk 6]
```diff
@@ -12,5 +12,9 @@
         step="1"
         [min]="minValue"
         [max]="maxValue"
+        #control="ngModel"
     />
+    <mat-error *ngIf="isRequired && control.hasError('required')">
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/number-input/number-input.component.html:17]:
```suggestion
    <mat-error *ngIf="control.hasError('required')">
```

File: src/lib/components/forms/form/form-components/select/select.component.html [hunk 7]
```diff
@@ -18,4 +19,7 @@
             {{ item.caption }}
         </mat-option>
     </mat-select>
+    <mat-error *ngIf="isRequired && control.hasError('required')">
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/select/select.component.html:22]:
```suggestion
    <mat-error *ngIf="control.hasError('required')">
```

File: src/lib/components/forms/form/form-components/slider/slider.component.html [hunk 8]
```diff
@@ -14,7 +14,11 @@
         [displayWith]="_precisionDisplayFn"
         (ngModelChange)="_onChangeCallback($event)"
         (blur)="_onTouchedCallback()"
+        #control="ngModel"
     >
     </mat-slider>
     <span *ngIf="showMinMax" class="align-self-center">{{ max }}</span>
 </div>
+<mat-error *ngIf="isRequired && control.hasError('required')">
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/slider/slider.component.html:22]:
```suggestion
<mat-error *ngIf="control.hasError('required')">
```

File: src/lib/components/forms/form/form-components/suggestion/suggestion.component.html [hunk 9]
```diff
@@ -29,4 +29,7 @@
             {{ option.caption }}
         </mat-option>
     </mat-autocomplete>
+    <mat-error *ngIf="isRequired && _control.hasError('required')">
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/suggestion/suggestion.component.html:32]:
```suggestion
    <mat-error *ngIf="_control.hasError('required')">
```

File: src/lib/components/forms/form/form-components/text-area/text-area.component.html [hunk 10]
```diff
@@ -4,15 +4,26 @@
...
         [maxlength]="maxLength"
+        [minlength]="minLength"
         [rows]="_maxRows"
         [required]="isRequired"
         autocomplete="off"
         [disabled]="isDisabled"
         [style.resize]="hasFixedHeight ? 'none' : 'vertical'"
     >
     </textarea>
+    <mat-error *ngIf="isRequired && control.hasError('required')">
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/text-area/text-area.component.html:20]:
```suggestion
    <mat-error *ngIf="control.hasError('required')">
```

File: src/lib/components/forms/form/form-components/text-area/text-area.component.ts [hunk 11]
```diff
@@ -46,10 +47,14 @@ export class TextAreaComponent implements ControlValueAccessor, TextAreaInterfac
     @Input()
     public tooltipPlacement: TerraPlacementEnum = TerraPlacementEnum.TOP;
 
-    /** Set a maximum number of characters allowed. */
+    /** Set a maximum number of required characters. */
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/text-area/text-area.component.ts:50]:
```suggestion
    /** Set a maximum number of allowed characters. */
```

File: src/lib/components/forms/form/form-components/text-input/text-input.component.html [hunk 12]
```diff
@@ -8,14 +8,31 @@
...
         [required]="isRequired"
         [disabled]="isDisabled"
         [readonly]="isReadonly"
-        [minLength]="minLength"
+        [minlength]="minLength"
         [maxlength]="maxLength"
         [iban]="isIban"
+        [email]="email"
     />
     <mat-error *ngIf="isIban && control.hasError('iban')">
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/text-input/text-input.component.html:20]:
```suggestion
    <mat-error *ngIf="control.hasError('iban')">
```

File: src/lib/components/forms/form/form-components/text-input/text-input.component.html [hunk 13]
```diff
@@ -8,14 +8,31 @@
...
+    <mat-error *ngIf="control.hasError('maxlength')">
+        {{ 'terraTextInput.maxLength' | translate: _locale.language:{ maxLength: maxLength } }}
+    </mat-error>
+    <mat-error *ngIf="control.hasError('minlength')">
+        {{ 'terraTextInput.minLength' | translate: _locale.language:{ minLength: minLength } }}
+    </mat-error>
+    <mat-error *ngIf="control.hasError('email')">
+        {{ 'terraTextInput.invalidEmail' | translate: _locale.language }}
+    </mat-error>
+    <mat-error *ngIf="isRequired && control.hasError('required')">
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/text-input/text-input.component.html:32]:
```suggestion
    <mat-error *ngIf="control.hasError('required')">
```

File: src/lib/components/forms/form/form-components/text-input/text-input.component.ts [hunk 14]
```diff
@@ -52,18 +52,22 @@ export class TextInputComponent implements ControlValueAccessor, TextInputInterf
     @Input()
     public isRequired: boolean = false;
 
-    /** @description Set the maximum number value allowed. */
+    /** @description Set the maximum number of required characters. */
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/text-input/text-input.component.ts:55]:
```suggestion
    /** @description Set the maximum number of allowed characters. */
```

File: src/lib/components/forms/form/form-components/text-input/text-input.component.ts [hunk 15]
```diff
@@ -52,18 +52,22 @@ export class TextInputComponent implements ControlValueAccessor, TextInputInterf
...
-    /** @description Set the minimum number value allowed. */
+    /** @description Set the minimum number of required characters. */
     @Input()
     public minLength: number;
 
     /** @description Set the label. */
     @Input()
     public name: string = '';
 
+    /** @description  whether entered text should be an email. **/
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/text-input/text-input.component.ts:67]:
```suggestion
    /** @description Whether entered text should be an email. **/
```

File: src/lib/components/forms/form/form-components/text-input/text-input.interface.ts [hunk 16]
```diff
@@ -6,4 +6,5 @@ export interface TextInputInterface extends TerraFormComponentBaseInterface {
     isReadonly: boolean;
     minLength: number;
     maxLength: number;
+    email: boolean;
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/text-input/text-input.interface.ts:9]:
```suggestion
    email: boolean;
    pattern: string;
```

File: src/lib/translations/l10n-terra-components.ts [hunk 17]
```diff
@@ -95,7 +95,10 @@ const de: {} = {
         }
     },
     terraTextInput: {
-        invalidIban: 'IBAN ist ungültig.'
+        invalidIban: 'IBAN ist ungültig.',
+        maxLength: 'Maximal {{maxLength}} Zeichen erlaubt.',
+        minLength: 'Mindestens {{minLength}} Zeichen benötigt.',
```
Comment (pweyrich) [inline src/lib/translations/l10n-terra-components.ts:100]: diese beiden sollten vllt zu der section mit validators verschoben werden. Sie gelten ja nicht nur für den textInput

File: src/lib/translations/l10n-terra-components.ts [hunk 18]
```diff
@@ -240,7 +249,10 @@ const en: {} = {
         }
     },
     terraTextInput: {
-        invalidIban: 'IBAN is not valid.'
+        invalidIban: 'IBAN is not valid.',
+        maxLength: 'A maximum of {{maxLength}} characters is allowed.',
+        minLength: 'A minimum of {{minLength}} characters is required.',
```
Comment (pweyrich) [inline src/lib/translations/l10n-terra-components.ts:254]: S.o.

File: src/lib/components/forms/form/form-components/text-input/text-input.component.spec.ts [hunk 19]
```diff
@@ -114,6 +114,40 @@ describe('TextInputComponent', () => {
...
+
+        expect(await formField.isControlValid()).toBeTruthy();
+
+        await input.setValue(invalidEmail);
+        fixture.detectChanges();
+
+        expect(await formField.isControlValid()).toBeFalsy();
+    });
+
+    it('should have the correct error message for Emails ', async () => {
```
Comment (johannabecker) [inline src/lib/components/forms/form/form-components/text-input/text-input.component.spec.ts:135]:
```suggestion
    it('should have the correct error message for emails ', async () => {
```

File: src/lib/components/forms/form/form-components/color-picker/color-picker.component.spec.ts [hunk 20]
```diff
@@ -111,4 +123,24 @@ describe('ColorPickerComponent', () => {
 
         expect(spy).toHaveBeenCalled();
     });
+
+    it('should display an error message when input does not match the hex color pattern', async () => {
```
Comment (johannabecker) [inline src/lib/components/forms/form/form-components/color-picker/color-picker.component.spec.ts:127]:
```suggestion
    it('should display an error message when input does not match the hex colour pattern', async () => {
```

**Review by johannabecker — COMMENTED** [review]
Ist mir nur beim Überfliegen aufgefallen. Vllt könnt ihr das noch mit anpassen.

File: src/lib/components/forms/form/form-components/text-area/text-area.component.spec.ts [hunk 21]
```diff
@@ -69,6 +81,66 @@ describe('TextAreaComponent', () => {
...
+        const formField: MatFormFieldHarness = await loader.getHarness(MatFormFieldHarness);
+        await input.setValue(validString);
+        expect(await formField.isControlValid()).toBe(true);
+        await input.setValue(invalidString);
+        expect(await formField.isControlValid()).toBe(false);
+    });
+
+    it('should have the correct error message for minLength ', async () => {
+        const minLength: number = 5;
+        const invalidString: string = Array(minLength - 1)
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/text-area/text-area.component.spec.ts:112]:
```suggestion
        const invalidString: string = new Array(minLength - 1)
```

File: src/lib/components/forms/form/form-components/text-area/text-area.component.spec.ts [hunk 22]
```diff
@@ -69,6 +81,66 @@ describe('TextAreaComponent', () => {
...
+        fixture.detectChanges();
+        expect(await formField.hasErrors()).toBeTrue();
+
+        const errors: Array<string> = await formField.getTextErrors();
+        expect(errors.includes('validators.minLength')).toBeTrue();
+    });
+
+    it('should have the correct error message for maxLength ', async () => {
+        const maxLength: number = 5;
+        const invalidString: string = Array(maxLength + 1)
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/text-area/text-area.component.spec.ts:129]:
```suggestion
        const invalidString: string = new Array(maxLength + 1)
```

File: src/lib/components/forms/form/form-components/text-input/text-input.component.spec.ts [hunk 23]
```diff
@@ -184,4 +218,91 @@ describe('TextInputComponent', () => {
...
+        const formField: MatFormFieldHarness = await loader.getHarness(MatFormFieldHarness);
+        await input.setValue(validString);
+        expect(await formField.isControlValid()).toBe(true);
+        await input.setValue(invalidString);
+        expect(await formField.isControlValid()).toBe(false);
+    });
+
+    it('should have the correct error message for minLength ', async () => {
+        const minLength: number = 5;
+        const invalidString: string = Array(minLength - 1)
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/text-input/text-input.component.spec.ts:277]:
```suggestion
        const invalidString: string = new Array(minLength - 1)
```

File: src/lib/components/forms/form/form-components/text-input/text-input.component.spec.ts [hunk 24]
```diff
@@ -184,4 +218,91 @@ describe('TextInputComponent', () => {
...
+        fixture.detectChanges();
+        expect(await formField.hasErrors()).toBeTrue();
+
+        const errors: Array<string> = await formField.getTextErrors();
+        expect(errors.includes('validators.minLength')).toBeTrue();
+    });
+
+    it('should have the correct error message for maxLength ', async () => {
+        const maxLength: number = 5;
+        const invalidString: string = Array(maxLength + 1)
```
Comment (pweyrich) [inline src/lib/components/forms/form/form-components/text-input/text-input.component.spec.ts:294]:
```suggestion
        const invalidString: string = new Array(maxLength + 1)
```

**Review by pweyrich — APPROVED** [review]
👍

**Review by niwolf — APPROVED** [review]

**Pull request merged**